packages/react-dom/src/client/ReactDOMComponent.js

```diff
@@ -197,6 +198,9 @@ if (__DEV__) {
     if (didWarnInvalidHydration) {
       return;
     }
+    if (clientValue && clientValue.$$typeof === REACT_OPAQUE_OBJECT_TYPE) {
```

Comment:
This seems like the wrong place to do this check. If we get here it's because we already thought that there was an error. If you do the check in `getValueForAttribute` instead, then we don't have to treat it as an error in the first place.